ssh/tailssh/tailssh.go

```diff
@@ -109,34 +109,87 @@ func (srv *server) OnPolicyChange() {
 	}
 }
 
-func (srv *server) newSSHServer() (*ssh.Server, error) {
-	ss := &ssh.Server{
+// conn represents a single SSH connection and it's associated
+// ssh.Server.
+type conn struct {
+	*ssh.Server
+	srv *server
+	ci  *sshConnInfo // set by NoClientAuthCallback
+
+	insecureSkipTailscaleAuth bool // used by tests.
+}
+
+func (c *conn) logf(format string, args ...any) {
+	if c.ci == nil {
+		c.srv.logf(format, args...)
+		return
+	}
+	format = fmt.Sprintf("%v: %v", c.ci.String(), format)
+	c.srv.logf(format, args...)
+}
+
+func (c *conn) PublicKeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
+	if c.ci == nil {
+		return false
+	}
+	if _, ok := c.srv.canProceed(c.ci, key); !ok {
+		c.logf("rejecting SSH public key %s", bytes.TrimSpace(gossh.MarshalAuthorizedKey(key)))
+		return false
+	}
+	c.logf("accepting SSH public key %s", bytes.TrimSpace(gossh.MarshalAuthorizedKey(key)))
+	return true
+}
+
+func (c *conn) NoClientAuthCallback(cm gossh.ConnMetadata) (*gossh.Permissions, error) {
+	if c.insecureSkipTailscaleAuth {
+		return nil, nil
+	}
+	ci, err := c.srv.connInfo(cm.User(), toIPPort(cm.LocalAddr()), toIPPort(cm.RemoteAddr()))
+	if err != nil {
+		c.logf("failed to get conninfo: %v", err)
+		return nil, gossh.ErrDenied
+	}
+	c.ci = ci
+	if _, ok := c.srv.canProceed(ci, nil /* without public key */); ok {
+		return nil, nil
+	}
+	if c.srv.havePubKeyPolicy(ci) {
+		return nil, fmt.Errorf("needs more info")
```

Comment:
return nil, errPubKeyRequired


var errPubKeyRequired = errors.New("some error") // to make NoClientAuthCallback proceed to public-key auth; not user visible